File: tests/human/x/pl/left_join_multi.pl
:- initialization(main).
customers([
  customer{id:1,name:'Alice'},
  customer{id:2,name:'Bob'}
]).
orders([
  order{id:100,customerId:1},
  order{id:101,customerId:2}
]).
items([
  item{orderId:100, sku:'a'}
]).
find_customer(Id, Cs, C) :- (member(X,Cs), X.id=Id -> C=X ; C=nil).
find_item(OrderId, Is, I) :- (member(X,Is), X.orderId=OrderId -> I=X ; I=nil).
left_join_multi([], _, _, []).
left_join_multi([O|Os], Cs, Is, [res{orderId:O.id,name:Name,item:I}|Rs]) :-
    find_customer(O.customerId,Cs,C),
    Name=C.name,
    find_item(O.id,Is,I),
    left_join_multi(Os, Cs, Is, Rs).
format_item(nil,'<nil>').
format_item(I,S) :- I \= nil, format(atom(S),'map[orderId:~w sku:~w]',[I.orderId,I.sku]).
main :-
    customers(Cs), orders(Os), items(Is),
    left_join_multi(Os, Cs, Is, Res),
    writeln('--- Left Join Multi ---'),
    forall(member(R,Res),(format_item(R.item,S),format('~w ~w ~w~n',[R.orderId,R.name,S]))),
    halt.

File: tests/human/x/java/LeftJoinMulti.java
import java.util.*;

class CustomerLJM {int id; String name; CustomerLJM(int id,String name){this.id=id;this.name=name;}}
class OrderLJM {int id; int customerId; OrderLJM(int id,int customerId){this.id=id;this.customerId=customerId;}}
class ItemLJM {int orderId; String sku; ItemLJM(int orderId,String sku){this.orderId=orderId;this.sku=sku;}}

public class LeftJoinMulti {
    public static void main(String[] args) {
        List<CustomerLJM> customers=Arrays.asList(
            new CustomerLJM(1,"Alice"),
            new CustomerLJM(2,"Bob")
        );
        List<OrderLJM> orders=Arrays.asList(
            new OrderLJM(100,1),
            new OrderLJM(101,2)
        );
        List<ItemLJM> items=Arrays.asList(
            new ItemLJM(100,"a")
        );
        System.out.println("--- Left Join Multi ---");
        for(OrderLJM o:orders){
            CustomerLJM c=null; for(CustomerLJM x:customers) if(x.id==o.customerId){c=x;break;}
            ItemLJM item=null; for(ItemLJM i:items) if(i.orderId==o.id){item=i;break;}
            System.out.println(o.id+" "+c.name+" "+(item!=null?item.sku:null));
        }
    }
}
